File: Novels/packages/6-mobile-android/StoryReader/app/src/main/java/com/storysphere/storyreader/repository/PaywallRepository.kt
package com.storysphere.storyreader.repository

import com.storysphere.storyreader.database.AppDatabase
import com.storysphere.storyreader.model.Chapter
import com.storysphere.storyreader.network.GraphQLService
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onStart
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PaywallRepository @Inject constructor(
    private val database: AppDatabase,
    private val graphQLService: GraphQLService
) {
    private val backgroundScope = CoroutineScope(SupervisorJob())
    
    data class PaywallInfo(
        val showPaywall: Boolean,
        val freeChaptersRemaining: Int,
        val chapterPrice: Int?,
        val isUnlocked: Boolean
    )
    
    fun getPaywallInfo(storyId: String, chapterId: String): Flow<PaywallInfo> {
        return database.chapterDao().getChapterFlow(chapterId)
            .map { entity ->
                PaywallInfo(
                    showPaywall = entity?.isPaid == true && entity.isUnlocked == false,
                    freeChaptersRemaining = 0, // TODO: Calculate from story metadata
                    chapterPrice = entity?.price,
                    isUnlocked = entity?.isUnlocked == true
                )
            }
            .onStart {
                backgroundScope.launch {
                    // Refresh chapter metadata from network
                    graphQLService.getChapter(chapterId).fold(
                        onSuccess = { chapter ->
                            database.chapterDao().insertChapter(chapter.toEntity())
                        },
                        onFailure = { }
                    )
                }
            }
    }
    
    suspend fun purchaseChapter(chapterId: String): Result<Chapter> {
        return graphQLService.purchaseChapter(chapterId).fold(
            onSuccess = { chapter ->
                database.chapterDao().insertChapter(chapter.toEntity())
                Result.success(chapter)
            },
            onFailure = { error -> Result.failure(error) }
        )
    }
    
    suspend fun purchaseBulk(chapterIds: List<String>): Result<List<Chapter>> {
        return graphQLService.purchaseBulk(chapterIds).fold(
            onSuccess = { chapters ->
                database.chapterDao().insertChapters(chapters.map { it.toEntity() })
                Result.success(chapters)
            },
            onFailure = { error -> Result.failure(error) }
        )
    }
    
    fun getPurchaseHistory(userId: String): Flow<List<Chapter>> {
        return database.chapterDao().getPurchasedChapters(userId)
            .map { entities -> entities.map { it.toDomain(null) } }
            .onStart {
                backgroundScope.launch {
                    graphQLService.getPurchaseHistory(userId).fold(
                        onSuccess = { chapters ->
                            database.chapterDao().insertChapters(chapters.map { it.toEntity() })
                        },
                        onFailure = { }
                    )
                }
            }
    }
}

File: Novels/packages/5-mobile-ios/StoryReader/Repository/PaywallRepository.swift
import Foundation
import Combine

// Paywall Repository - Paywall logic
class PaywallRepository {
    private let graphQLService = GraphQLService.shared
    
    func getPaywallInfo(storyId: String, userId: String) -> AnyPublisher<PaywallViewModel.PaywallInfo, Error> {
        return Future { [weak self] promise in
            guard let self = self else {
                promise(.failure(RepositoryError.unknown))
                return
            }
            
            Task {
                do {
                    let info = try await self.graphQLService.getPaywallInfo(storyId: storyId, userId: userId)
                    promise(.success(info))
                } catch {
                    promise(.failure(error))
                }
            }
        }
        .eraseToAnyPublisher()
    }
    
    func purchaseChapter(chapterId: String, userId: String) -> AnyPublisher<Void, Error> {
        return Future { [weak self] promise in
            guard let self = self else {
                promise(.failure(RepositoryError.unknown))
                return
            }
            
            Task {
                do {
                    try await self.graphQLService.purchaseChapter(chapterId: chapterId, userId: userId)
                    promise(.success(()))
                } catch {
                    promise(.failure(error))
                }
            }
        }
        .eraseToAnyPublisher()
    }
    
    func purchaseBulk(chapterIds: [String], userId: String) -> AnyPublisher<Void, Error> {
        return Future { [weak self] promise in
            guard let self = self else {
                promise(.failure(RepositoryError.unknown))
                return
            }
            
            Task {
                do {
                    try await self.graphQLService.purchaseBulk(chapterIds: chapterIds, userId: userId)
                    promise(.success(()))
                } catch {
                    promise(.failure(error))
                }
            }
        }
        .eraseToAnyPublisher()
    }
}


